### PR TITLE
upgrade to pytorch 0.4.0

### PR DESCRIPTION
**Pull request by joelgrus**
I think you will be pleased with the results; on the whole the codebase feels quite a bit cleaner.

For the most part I tried to follow the migration guide:

http://pytorch.org/2018/04/22/0_4_0-migration-guide.html

at a high level that included the following

* get rid of every reference to `torch.autograd.Variable`
* replace the typechecks for tensors with their recommended (unidiomatic way), and factor out into a `is_tensor` function
* dealing with the new 0-tensors everywhere (.item() instead of [0])
* replacing the `volatile` flags on tensors with `with torch.no_grad()` context blocks (I *think* I put them in the right places)
* all the `initializer` functions are now `initializer_`
* I had to upgrade tensorboardX to 1.2 (the version we were on didn't work with pytorch 0.4)
* replacing the clunky `new()` logic for making new tensors on specific devices with the new pytorch 0.4 ways of doing things `old_tensor.new_zeros(10)`, `old_tensor.new_tensor([x, y, z])` and so on

what I did not do:

* any of the "writing device agnostic code" stuff in that guide. that's likely a good idea, but it deserves its own PR.

--

When I started there were 100s of warnings; there are still two left that I'm not 100% sure how to fix:

```
tests/models/sniff_test.py::SniffTest::test_coreference_resolution
  /home/joelg/miniconda3/envs/pytorch40/lib/python3.6/site-packages/torch/nn/modules/rnn.py:38: UserWarning: dropout option adds dropout after all but last recurrent layer, so non-zero dropout expects num_layers greater than 1, but got dropout=0.2 and num_layers=1
    "num_layers={}".format(dropout, num_layers))

tests/models/sniff_test.py::SniffTest::test_machine_comprehension
  /home/joelg/miniconda3/envs/pytorch40/lib/python3.6/site-packages/torch/nn/modules/rnn.py:38: UserWarning: dropout option adds dropout after all but last recurrent layer, so non-zero dropout expects num_layers greater than 1, but got dropout=0.2 and num_layers=1
    "num_layers={}".format(dropout, num_layers))
```

both only occur in the `SniffTest`, so maybe they are not worth worrying about. or maybe we need to update the models that those two rely on? in any case, I'm not losing too much sleep over them.

### CODE REVIEW

File: allennlp/nn/decoding/decoder_trainers/expected_risk_minimization.py [hunk 1]
```diff
@@ -49,7 +49,11 @@ def decode(self,
             # Finished model scores are log-probabilities of the predicted sequences. We convert
             # log probabilities into probabilities and re-normalize them to compute expected cost under
             # the distribution approximated by the beam search.
-            costs = torch.cat(finished_costs[batch_index])
+
+            # finished_costs[batch_index] could be a list of 0-tensors or 1-tensors. We use .view(-1)
+            # to make sure they're treated as 1-tensors and then concatenated appropriately.
+            # TODO(joelgrus): make sure this is correct
```
Comment (matt-gardner) [inline allennlp/nn/decoding/decoder_trainers/expected_risk_minimization.py]: This looks like it shouldn't be right to me.  Well, it looks like it probably does the right thing, but I think fixing this where the `finished_costs` gets created is the right solution, instead of this one.  My guess is that something is getting created as a scalar now, where we really want it as a 1-dim tensor.

**Comment by schmmd** [comment]
Well, I only got up to `allennlp/modules/augmented_lstm.py` but what I looked at was great!  Have you tried our pre-trained models on this branch?  Thanks for putting this together.

**Comment by joelgrus** [comment]
only via the `SniffTest`, which passes.

off the top of my head, there's no reason why *these changes* should affect existing trained models (unless I did something incorrectly). but it's possible there are other changes in pytorch 0.4 that would.

File: allennlp/nn/decoding/decoder_trainers/expected_risk_minimization.py [hunk 2]
```diff
@@ -56,8 +55,10 @@ def decode(self,
             # Finished model scores are log-probabilities of the predicted sequences. We convert
             # log probabilities into probabilities and re-normalize them to compute expected cost under
             # the distribution approximated by the beam search.
-            costs = torch.cat(finished_costs[batch_index])
-            logprobs = torch.cat(finished_model_scores[batch_index])
+
+            print("size:", finished_costs[batch_index])
```
Comment (DeNeutoy) [inline allennlp/nn/decoding/decoder_trainers/expected_risk_minimization.py]: remove print statement

File: allennlp/nn/decoding/decoder_trainers/maximum_marginal_likelihood.py [hunk 3]
```diff
@@ -65,7 +64,8 @@ def decode(self,
         batch_scores = self._group_scores_by_batch(finished_states)
         loss = 0
         for scores in batch_scores.values():  # we don't care about the batch index, just the scores
-            loss += -util.logsumexp(torch.cat(scores))
+            print("scores", scores)
```
Comment (DeNeutoy) [inline allennlp/nn/decoding/decoder_trainers/maximum_marginal_likelihood.py]: remove print statement

File: allennlp/nn/decoding/decoder_trainers/maximum_marginal_likelihood.py [hunk 4]
```diff
@@ -65,7 +64,8 @@ def decode(self,
...
+            loss += -util.logsumexp(torch.cat([tensor.view(-1) for tensor in scores]))
```
Comment (DeNeutoy) [inline allennlp/nn/decoding/decoder_trainers/maximum_marginal_likelihood.py]: I think doing a squeeze here is tighter - I was confused when I read this, as it isn't immediately obvious that it's a 1D tensor and that doing a view like this is the same.

File: allennlp/nn/decoding/decoder_trainers/expected_risk_minimization.py [hunk 5]
```diff
@@ -56,8 +55,10 @@ def decode(self,
...
+            costs = torch.cat([tensor.view(-1) for tensor in finished_costs[batch_index]])
+            logprobs = torch.cat([tensor.view(-1) for tensor in finished_model_scores[batch_index]])
```
Comment (DeNeutoy) [inline allennlp/nn/decoding/decoder_trainers/expected_risk_minimization.py]: Can you use `squeeze` instead of `view` here?

**Comment by schmmd** [comment]
[MC model](http://beaker.allenai.org/ex/ex_xbrfg02q5nr7/tasks) evaluation matches.
[TE model](http://beaker.allenai.org/ex/ex_s9f5kgm4nboe/tasks) evaluation matches.

File: allennlp/training/metrics/boolean_accuracy.py [hunk 6]
```diff
@@ -34,7 +34,6 @@ def __call__(self,
         mask: ``torch.Tensor``, optional (default = None).
             A tensor of the same shape as ``predictions``.
         """
-        # Get the data from the Variables.
         predictions, gold_labels, mask = self.unwrap_to_tensors(predictions, gold_labels, mask)
```
Comment (matt-gardner) [inline allennlp/training/metrics/boolean_accuracy.py]: This method feels like it should be renamed to something like `detach_and_move_to_cpu` (or something similar but more compactly named...), and maybe just moved to `nn.util`.

File: allennlp/training/metrics/metric.py [hunk 7]
```diff
@@ -50,9 +51,11 @@ def from_params(cls, params: Params, vocab: Optional[Vocabulary] = None):
...
     def unwrap_to_tensors(*tensors):
         """
-        If you actually passed in Variables to a Metric instead of Tensors, there will be
+        If you actually passed gradient-tracking Tensors to a Metric, there will be
         a huge memory leak, because it will prevent garbage collection for the computation
         graph. This method ensures that you're using tensors directly and that they are on
         the CPU.
         """
-        return (x.data.cpu() if isinstance(x, torch.autograd.Variable) else x for x in tensors)
+        # pylint: disable=unidiomatic-typecheck
```
Comment (matt-gardner) [inline allennlp/training/metrics/metric.py]: I don't think you need this anymore.

File: allennlp/training/metrics/metric.py [hunk 8]
```diff
@@ -50,9 +51,11 @@ def from_params(cls, params: Params, vocab: Optional[Vocabulary] = None):
...
+        return (x.data.cpu() if is_tensor(x) else x
```
Comment (matt-gardner) [inline allennlp/training/metrics/metric.py]: One line?

Comment (matt-gardner) [reply allennlp/training/metrics/metric.py]: And probably just fail if you pass something that's not a tensor, instead of checking `is_tensor`.

Comment (matt-gardner) [reply allennlp/training/metrics/metric.py]: `.detach()` instead of `.data`

File: allennlp/training/metrics/metric.py [hunk 9]
```diff
@@ -50,9 +51,11 @@ def from_params(cls, params: Params, vocab: Optional[Vocabulary] = None):
     @staticmethod
     def unwrap_to_tensors(*tensors):
```
Comment (matt-gardner) [inline allennlp/training/metrics/metric.py]: Can we give a type annotation here now?  Not sure how that works with `*args`, though...

File: allennlp/training/trainer.py [hunk 10]
```diff
@@ -463,11 +461,11 @@ def _train_epoch(self, epoch: int) -> Dict[str, float]:
                 # get the magnitude of parameter updates for logging
                 # We need a copy of current parameters to compute magnitude of updates,
                 # and copy them to CPU so large models won't go OOM on the GPU.
-                param_updates = {name: param.detach().data.cpu().clone()
+                param_updates = {name: param.data.cpu().clone()
```
Comment (matt-gardner) [inline allennlp/training/trainer.py]: `.detach()` instead of `.data`

File: allennlp/training/trainer.py [hunk 11]
```diff
@@ -463,11 +461,11 @@ def _train_epoch(self, epoch: int) -> Dict[str, float]:
...
                                  for name, param in self._model.named_parameters()}
                 self._optimizer.step()
                 for name, param in self._model.named_parameters():
-                    param_updates[name].sub_(param.detach().data.cpu())
+                    param_updates[name].sub_(param.data.cpu())
```
Comment (matt-gardner) [inline allennlp/training/trainer.py]: `.detach()` instead of `.data`

File: tests/nn/util_test.py [hunk 12]
```diff
@@ -530,7 +513,7 @@ def test_flatten_and_batch_shift_indices(self):
                                [[2, 1, 0, 7],
                                 [7, 7, 2, 3],
                                 [0, 0, 4, 2]]])
-        indices = Variable(torch.LongTensor(indices))
+        indices = torch.from_numpy(indices).long()
```
Comment (matt-gardner) [inline tests/nn/util_test.py]: `dtype=torch.long`

File: tests/nn/util_test.py [hunk 13]
```diff
@@ -546,8 +529,7 @@ def test_batched_index_select(self):
         targets = torch.ones([2, 10, 3]).cumsum(1) - 1
         # Make the second batch double it's index so they're different.
         targets[1, :, :] *= 2
-        indices = Variable(torch.LongTensor(indices))
-        targets = Variable(targets)
+        indices = torch.from_numpy(indices).long()
```
Comment (matt-gardner) [inline tests/nn/util_test.py]: `dtype=torch.long`

File: tests/nn/util_test.py [hunk 14]
```diff
@@ -568,8 +550,7 @@ def test_flattened_index_select(self):
         targets = torch.ones([2, 6, 3]).cumsum(1) - 1
         # Make the second batch double it's index so they're different.
         targets[1, :, :] *= 2
-        indices = Variable(torch.LongTensor(indices))
-        targets = Variable(targets)
+        indices = torch.from_numpy(indices).long()
```
Comment (matt-gardner) [inline tests/nn/util_test.py]: `dtype=torch.long`

File: tests/training/trainer_test.py [hunk 15]
```diff
@@ -244,5 +244,5 @@ def test_sparse_clip_grad(self):
         # Now try to clip the gradients.
         _ = sparse_clip_norm([embedding.weight], 1.5)
         # Final norm should be 1.5
-        grad = embedding.weight.grad.data.coalesce()
-        self.assertAlmostEqual(grad._values().norm(2.0), 1.5, places=5) # pylint: disable=protected-access
+        grad = embedding.weight.grad.data.coalesce()  # pylint: disable=no-member
```
Comment (matt-gardner) [inline tests/training/trainer_test.py]: Do you need the `.data` here still?

**Comment by schmmd** [comment]
BiDAF trained too: http://beaker.allenai.org/ex/ex_fh0xt1u5ytst/tasks

File: allennlp/nn/decoding/decoder_trainers/maximum_marginal_likelihood.py (hunk 4 above)
Comment (joelgrus) [reply allennlp/nn/decoding/decoder_trainers/maximum_marginal_likelihood.py]: the actual issue here (which I didn't dig into too deeply) is that sometimes the score was a 0-tensor and sometimes it was a 1-tensor, and the `.view()` is ensuring it's a 1-tensor.

File: allennlp/common/util.py [hunk 16]
```diff
@@ -47,9 +52,7 @@ def sanitize(x: Any) -> Any:  # pylint: disable=invalid-name,too-many-return-sta
     if isinstance(x, (str, float, int, bool)):
         # x is already serializable
         return x
-    elif isinstance(x, torch.autograd.Variable):
-        return sanitize(x.data)
-    elif isinstance(x, torch._TensorBase):  # pylint: disable=protected-access
+    elif is_tensor(x):
         # tensor needs to be converted to a list (and moved to cpu if necessary)
         return x.cpu().tolist()
```
[the remark this answers is not in the document]
Comment (joelgrus) [reply allennlp/common/util.py]: I don't think it would do anything. as I understand it, the main reason to use `.data` is to "break free" of the computational graph; once you convert to a list you're not part of the computational graph anyway

File: allennlp/models/semantic_parsing/wikitables/wikitables_semantic_parser.py [hunk 17]
```diff
@@ -398,7 +397,7 @@ def forward(self,  # type: ignore
                 if i in best_final_states:
                     best_action_indices = best_final_states[i][0].action_history[0]
                     if target_action_sequences is not None:
-                        # Use a Tensor, not a Variable, to avoid a memory leak.
+                        # Detach to avoid a memory leak.
                         targets = target_action_sequences[i].data
```
[the remark this answers is not in the document]
Comment (joelgrus) [reply allennlp/models/semantic_parsing/wikitables/wikitables_semantic_parser.py]: I was a little confused on this, here's what the migration docs say:

```
However, .data can be unsafe in some cases. Any changes on x.data wouldn’t be tracked by autograd, and the computed gradients would be incorrect if x is needed in a backward pass. A safer alternative is to use x.detach(), which also returns a Tensor that shares data with requires_grad=False, but will have its in-place changes reported by autograd if x is needed in backward.
```

to me that suggests that `detach` still retains some connection to the computational graph and might still result in a memory leak?

File: allennlp/modules/encoder_base.py [hunk 18]
```diff
@@ -235,8 +233,7 @@ def _update_states(self,
         if self._states is None:
             # We don't already have states, so just set the
             # ones we receive to be the current state.
-            self._states = tuple([torch.autograd.Variable(state.data)
-                                  for state in new_unsorted_states])
+            self._states = tuple(new_unsorted_states)
```
[the remark this answers is not in the document]
Comment (joelgrus) [reply allennlp/modules/encoder_base.py]: oops, just a mistake on my part.

File: allennlp/nn/util.py [hunk 19]
```diff
@@ -345,12 +335,13 @@ def get_text_field_mask(text_field_tensors: Dict[str, torch.Tensor],
     the mask.  Most frequently this will be a character id tensor, but it could also be a
     featurized representation of each token, etc.
 
+    TODO(joelgrus): can we change this?
     NOTE: Our functions for generating masks create torch.LongTensors, because using
-    torch.ByteTensors inside Variables makes it easy to run into overflow errors
+    torch.ByteTensors  makes it easy to run into overflow errors
```
[the remark this answers is not in the document]
Comment (joelgrus) [reply allennlp/nn/util.py]: I need to fix this whole comment, because I'm not sure if it's true anymore. Now that variables don't exist, is this still a real problem? @DeNeutoy

File: allennlp/training/metrics/metric.py (hunk 8 above)
Comment (joelgrus) [reply allennlp/training/metrics/metric.py]: removing the `else x` behavior breaks a lot of tests (not sure if it's just how the tests are written or if it actually breaks our code)

Comment (matt-gardner) [reply allennlp/training/metrics/metric.py]: I just checked this, and it's because we pass in masks that could be `None`.  It might be worth changing this to `(x.data.cpu() if x is not None else None for x in tensors)`, just to fail explicitly instead of silently if you're passing something in that we're not expecting.  Not a big deal, though.

File: allennlp/models/semantic_parsing/wikitables/wikitables_semantic_parser.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (matt-gardner) [reply allennlp/models/semantic_parsing/wikitables/wikitables_semantic_parser.py]: I just asked about this in the pytorch slack, and I got this response:

"Variables returned by `.detach()` do not hold on to the computation graph. You can almost certainly use `.detach()` in places where you are using `.data`

"They do share a version counter with the original tensor. This doesn't require any extra memory usage. It just means that changes to the detached variable increment the (shared) version counter on  original tensor. So if you do backprop on the original tensor, and it's needed for a gradient computation, you would get an error message instead of just incorrect gradients."

**Comment by murphp15** [comment]
@schmmd On a number of the PR's I have noticed that you guys post links to beaker. 
Is this something the public should not have access to?

**Comment by matt-gardner** [comment]
@murphp15, beaker is an experiment management system that we're building internally at AI2.  We submit jobs to a cluster using beaker, and it keeps track of the experiments that we run.  It's pretty nice.  The plan is to make it open source eventually, but I don't know what the timeline is for that.

**Comment by schmmd** [comment]
@murphp15 it's internal, but hopefully we'll be able to share external links someday.  We're working out our plan on that currently.

**Comment by schmmd** [comment]
We plan to merge this on Tuesday.
